Objects/stringlib/fastsearch.h

```diff
@@ -324,224 +335,350 @@ typedef struct STRINGLIB(_pre) {
-                         Py_ssize_t len_needle)
+STRINGLIB(two_way_find)(const STRINGLIB_CHAR *haystack,
+                        Py_ssize_t len_haystack,
+                        const STRINGLIB_CHAR *needle,
+                        Py_ssize_t len_needle,
+                        Py_ssize_t maxcount,
+                        int mode)
 {
-    LOG("###### Finding \"%s\" in \"%s\".\n", needle, haystack);
-    STRINGLIB(prework) p;
-    STRINGLIB(_preprocess)(needle, len_needle, &p);
-    return STRINGLIB(_two_way)(haystack, len_haystack, &p);
+    STRINGLIB(prework) pw;
+    STRINGLIB(_preprocess)(needle, len_needle, &pw, 1, 1);
+    return STRINGLIB(_two_way)(haystack, len_haystack, maxcount, mode, &pw);
 }
 
 
 static Py_ssize_t
-STRINGLIB(_two_way_count)(const STRINGLIB_CHAR *haystack,
-                          Py_ssize_t len_haystack,
-                          const STRINGLIB_CHAR *needle,
-                          Py_ssize_t len_needle,
-                          Py_ssize_t maxcount)
+STRINGLIB(horspool_find)(const STRINGLIB_CHAR* s, Py_ssize_t n,
+                        const STRINGLIB_CHAR* p, Py_ssize_t m,
+                        Py_ssize_t maxcount, int mode, int dynamic)
 {
-    LOG("###### Counting \"%s\" in \"%s\".\n", needle, haystack);
-    STRINGLIB(prework) p;
-    STRINGLIB(_preprocess)(needle, len_needle, &p);
-    Py_ssize_t index = 0, count = 0;
-    while (1) {
-        Py_ssize_t result;
-        result = STRINGLIB(_two_way)(haystack + index,
-                                     len_haystack - index, &p);
-        if (result == -1) {
-            return count;
-        }
-        count++;
-        if (count == maxcount) {
-            return maxcount;
-        }
-        index += result + len_needle;
+    /* Boyer–Moore–Horspool algorithm
+       with optional dynamic fallback to Two-Way algorithm */
+    Py_ssize_t mlast = m - 1;
+    const Py_ssize_t width = n - m;
+    const STRINGLIB_CHAR plast = p[mlast];
+    const STRINGLIB_CHAR *ss = s + mlast;   // Last window character
+    STRINGLIB_CHAR slast;
+
+    // Pre-Work
+    STRINGLIB(prework) pw;
+    STRINGLIB(_preprocess)(p, m, &pw, 0, 1);
+    Py_ssize_t gap = pw.gap;
+    SHIFT_TYPE *table = pw.table;
+
+    // Use Bloom for len(needle) <= 64
+    // Initialization is costly for long needles
+    // And this is not much beneficial for large set(needle)
+    unsigned long mask = 0;
+    Py_ssize_t bloom_gap = 0;
+    if (m <= 64) {
+        LOG("Using Bloom skip\n");
+        // Note: bloom_gap("___aa") = 1
+        bloom_gap = m;
+        STRINGLIB_BLOOM_ADD(mask, plast);
+        for (Py_ssize_t i = 0; i < mlast; i++) {
+            STRINGLIB_BLOOM_ADD(mask, p[i]);
+            if (p[i] == plast) {
+                bloom_gap = mlast - i;
+            }
+        }
     }
-    return count;
+    // Horspool Calibration
+    const float hrs_lcost = 4.0;        // average loop cost
+    const float hrs_hcost = 0.4;        // false positive hit cost
```

Comment:
You're storing a double in a float. 2 solutions:

```suggestion
    const float hrs_lcost = 4.0f;        // average loop cost
    const float hrs_hcost = 0.4f;        // false positive hit cost
```

OR

```suggestion
    const double hrs_lcost = 4.0;        // average loop cost
    const double hrs_hcost = 0.4;        // false positive hit cost
```